src/server/pfs/server/driver.go

```diff
@@ -1407,64 +1407,199 @@ func (d *driver) deleteCommit(ctx context.Context, commit *pfs.Commit) error {
-		if branchInfo.Head != nil && branchInfo.Head.ID == commitInfo.Commit.ID {
-			if err := d.createBranch(ctx, branchInfo.Branch, commitInfo.ParentCommit, nil); err != nil {
+			// Update repo size
+			// TODO this is basically wrong. Other commits could share data with the
+			// commit being removed, in which case we're subtracting too much. Doing
+			// the right thing is hard, though. Hopefully nobody ends up with a
+			// negative repo size and cares too much
+			repoInfo := &pfs.RepoInfo{}
+			if err := repos.Get(commit.Repo.Name, repoInfo); err != nil {
+				return err
+			}
+			repoInfo.SizeBytes -= commitInfo.SizeBytes
+			return repos.Put(commit.Repo.Name, repoInfo)
+		}
+
+		// Delete the commit itself and subtract the size of the commit from repo
+		// size.
+		deleteCommit(commits, commit)
+
+		// Delete all of the downstream commits of 'commit'
+		for _, subvCommit := range commitInfo.Subvenance {
+			downstreamRepo := d.commits(subvCommit.Repo.Name).ReadWrite(stm)
+			deleteCommit(downstreamRepo, subvCommit)
+		}
+
+		// Remove the commits in 'deleted' from all upstream commits' subvenance
+		for _, provCommit := range commitInfo.Provenance {
+			provCI := &pfs.CommitInfo{}
+			upstreamRepo := d.commits(provCommit.Repo.Name).ReadWrite(stm)
+			if err := upstreamRepo.Get(provCommit.ID, provCI); err != nil {
+				return err
+			}
+			// Remove deleted commits from provCI.Subvenance (copy 'from' -> 'to')
+			to := 0
+			for from, c := range provCI.Subvenance {
+				if _, ok := deleted[c.ID]; ok {
+					continue
+				}
+				provCI.Subvenance[to] = provCI.Subvenance[from]
+				to++
+			}
+			provCI.Subvenance = provCI.Subvenance[:to]
+			// Write updated commitInfo back to etcd
+			if err := upstreamRepo.Put(provCommit.ID, provCI); err != nil {
 				return err
 			}
 		}
+		return nil
+	}); err != nil {
+		return err
 	}
 
-	// Delete the commit itself and subtract the size of the commit
-	// from repo size.
-	_, err = col.NewSTM(ctx, d.etcdClient, func(stm col.STM) error {
-		repos := d.repos.ReadWrite(stm)
-		repoInfo := new(pfs.RepoInfo)
-		if err := repos.Get(commit.Repo.Name, repoInfo); err != nil {
+	if deleteScratch {
+		// Delete the scratch space for this commit
+		// TODO put scratch spaces in a collection and do this in the txn above
+		prefix, err := d.scratchCommitPrefix(ctx, commit)
+		if err != nil {
+			return err
+		}
+		if _, err = d.etcdClient.Delete(ctx, prefix, etcd.WithPrefix()); err != nil {
 			return err
 		}
-		repoInfo.SizeBytes -= commitInfo.SizeBytes
-		repos.Put(commit.Repo.Name, repoInfo)
+	}
 
-		commits := d.commits(commit.Repo.Name).ReadWrite(stm)
-		return commits.Delete(commit.ID)
-	})
+	// Rewrite all commits in affected repos so that:
+	// - No commit's parent is a deleted commit.
+	// - No branch points to a deleted commit
+	// TODO this should happen in the txn as well, but can't because we have to
+	// list branches and commits in 'modifiedRepos', and you can't list in a txn
```

Comment:
So we could in theory fix this by having commits store references to their parents and the branches they're currently the head of right? That seems like it might be worth doing in the future but probably not as part of this release.

Reply:
I believe so. For branches, you could even have RepoInfos store the list of branches in the repo, so that you can retrieve the list without calling `col.List()`. But for commits I don't see an alternative to child pointers.

Reply:
Yup, I think so. I originally thought of putting the list of branches in RepoInfo, so that you could retrieve the list without calling `col.List()`, but a) that might perform badly for the spec repo, which has a large number of branches, and b) doesn't help with rewriting `ParentCommit`.